WIP: KNMI Climate Scenarios recipe

evertrol opened the pull request:
This pull request is a first attempt at implementing a recipe that runs the first step in the KNMI Climate Scenarios (kcs), as detailed in [Lenderink et al, 2014](https://iopscience.iop.org/article/10.1088/1748-9326/9/11/115008). This PR is very much still work in progress.

A reference implementation exists at https://github.com/eucp-project/kcs, and the current implementation follows that closely. This recipe still misses the various plots that are produced: it only does the essential of resampling the model of interest.

Things to look at and possibly (probably) change:

- I've structured the recipe with three diagnostic scripts. The scripts, however, are interdepent, and essentially one script: output produced by one script is required input of the next script (as indicated by 'ancestors').
On the other hand, there is some practical use and reasoning structuring it this way: it provides the substeps as detailed in Lenderink et al, which are relatively separate.

Perhaps the structure should be different, but this also relates to the next two points.

- The second and third diagnostic script need output from their ancestor, which is a simple CSV file. There is now some odd semi-hardcoding that looks for e.g. `../../diagnostic1/script2/<filename>.csv. Is there a better to provide input (and output!) files, in particular in a more fixed location than a date-generated output directory?

- For the second diagnostic, only one model is needed, with a different preprocessor than in the first diagnostic. Is there a way to select only one model to be run through the second preprocessor? Currently, all datasets are also ran through the second preprocessor. (Note: when plots get implemented, this may be necessary anyway.)

Further points:

- I've used a small (CMIP 5) example for now, with the CCSM4 as the model of interest. The actual model is EC-EARTH in the original implementation and in Lenderink et al, but the current EC-EARTH data in the CMIP 5 archive appears to have too many problems to be usable.
  The full CMIP 5 dataset should be used for the initial step, but this obviously takes too long during testing.
  Lines starting with just `##%` should be included in the final recipe, that is, removing `##%` will set this right.

- A seemingly odd thing that happened when I tested this recipe on a local machine, is that it crashed for a slightly larger dataset, for `diagnostic2/script1`. It looks like a out-of-memory issue due to multiprocessing overhead, since the traceback shows the multiprocessing module being used. I had already removed all multiprocessing from this script before starting with this recipe, and I have as of yet no idea what is eating up all the memory, nor where multiprocessing comes into play.


Other improvements to do are:

- refactoring
  There is probably still quite a bit of code that can be refactored, or generally be made more efficient. Part of this is a remnant from the reference implementation, from which a lot is copy-pasted.

- linting
  I have not run pylint or flake8, or did any clean up. (This includes using a better logger name, for example.)

----------

**Tasks**

-   [ ] [Create an issue](https://github.com/ESMValGroup/ESMValTool/issues) to discuss what you are going to do, if you haven't done so already (and add the link at the bottom)
-   [x] Give this pull request a descriptive title that can be used as a one line summary in a changelog
-   [ ] Make sure your code is composed of functions of no more than 50 lines and uses meaningful names for variables
-   [ ] Circle/CI tests pass. Status can be seen below your pull request. If the tests are failing, click the link to find out why.
-   [ ] Preferably Codacy code quality checks pass, however a few remaining hard to solve Codacy issues are still acceptable. Status can be seen below your pull request. If there is an error, click the link to find out why. If you suspect Codacy may be wrong, please ask by commenting.
-   [x] Please use `yamllint` to check that your YAML files do not contain mistakes 
-   [x] (Only if really necessary) Add any additional dependencies needed for the diagnostic script to setup.py, esmvaltool/install/R/r_requirements.txt or esmvaltool/install/Julia/julia_requirements.txt (depending on the language of your script) and also to meta.yaml for conda dependencies (includes Python and others, but not R/Julia)
-   [x] If new dependencies are introduced, check that the license is compatible with [Apache2.0](https://github.com/ESMValGroup/ESMValTool/blob/master/LICENSE)

New recipe/diagnostic

-   [ ] Add documentation for the recipe to the `doc/sphinx/source/recipes` folder and add a new entry to index.rst
-   [ ] Add provenance information

* * *

bouweandela commented:
Hi Evert,

Great to see a first EUCP pull request coming in!

Could you please use the relevant check boxes from the [pull request template](https://github.com/ESMValGroup/ESMValTool/blob/master/.github/pull_request_template.md)? You can delete or cross out anything that doesn't apply. It makes it much easier to communicate about what needs to be done before a pull request can be merged and easier for reviewers to see what the current status of a pull request is.

evertrol commented:
@bouweandela Apologies; I had removed them, since there was quite a bit of text in the PR. I've added them back at the bottom.
I'm not sure what to do with the first item: I've gotten used to the fact that PRs often are treated as issues, and for a totally new recipe, I'm not sure if creating a separate issue applies.

bouweandela commented:
> I'm not sure if creating a separate issue applies.

Issues are a good place to discuss things, so I would say yes. If you start discussing things in pull requests, they get so long that it becomes impossible to keep track of what is going on. Especially when discussion takes place both in comments on lines of code and in comments below the pull request.


>  Is there a better to provide input (and output!) files, in particular in a more fixed location than a date-generated output directory?

Yes, there is the location defined in config-user.yml: https://github.com/ESMValGroup/ESMValTool/blob/fef8ef696d17c3b57b53559b46f223939a084565/config-user-example.yml#L37-L38
that is passed on to the diagnostic script.

> For the second diagnostic, only one model is needed, with a different preprocessor than in the first diagnostic. Is there a way to select only one model to be run through the second preprocessor? 

Yes, use the `additional_datasets` key to define datasets per diagnostic, instead of the global `datasets` definition that applies to the entire recipe, see also:
https://esmvaltool.readthedocs.io/projects/esmvalcore/en/latest/recipe/overview.html#diagnostic-and-variable-specific-datasets


> It looks like a out-of-memory issue due to multiprocessing overhead, since the traceback shows the multiprocessing module being used. I had already removed all multiprocessing from this script before starting with this recipe, and I have as of yet no idea what is eating up all the memory, nor where multiprocessing comes into play.

The recipe defines a number of tasks, as described [here](https://esmvaltool.readthedocs.io/projects/esmvalcore/en/latest/recipe/overview.html#the-diagnostics-section-defines-tasks). By default these tasks are executed in parallel using the `multiprocessing` module. https://github.com/ESMValGroup/ESMValTool/blob/fef8ef696d17c3b57b53559b46f223939a084565/config-user-example.yml#L25-L32
Change the setting above to disable this


> This includes using a better logger name, for example

`logger` is a great name for a logger (much more readable than LOGGER) so I've configured our `pylint` such that it accepts this as a good name:
https://github.com/ESMValGroup/ESMValTool/blob/fef8ef696d17c3b57b53559b46f223939a084565/.pylintrc#L196-L197

bouweandela commented:
> I've added them back at the bottom.

Thanks!

Peter9192 commented:
Work is continued in #1667
